plutus-ledger-api/src/PlutusLedgerApi/Common/Versions.hs

```diff
@@ -153,44 +104,173 @@ ledgerLanguageIntroducedIn = \case
+  , MkCons, HeadList, TailList, NullList, ChooseData, ConstrData, MapData, ListData
+  , IData, BData, UnConstrData, UnMapData, UnListData, UnIData, UnBData, EqualsData
+  , MkPairData, MkNilData, MkNilPairData
+  ]
+
+batch2 :: [DefaultFun]
+batch2 =
+  [ SerialiseData ]
+
+batch3 :: [DefaultFun]
+batch3 =
+  [ VerifyEcdsaSecp256k1Signature, VerifySchnorrSecp256k1Signature ]
+
+-- `cekCase` and `cekConstr` costs come between Batch 3 and Batch 4 in the
+-- PlutusV3 cost model parameters, although that's irrelevant here.
+
+-- batch4, excluding IntegerToByteString and ByteStringToInteger.
+batch4a :: [DefaultFun]
+batch4a =
+  [ Bls12_381_G1_add, Bls12_381_G1_neg, Bls12_381_G1_scalarMul
+  , Bls12_381_G1_equal, Bls12_381_G1_hashToGroup
+  , Bls12_381_G1_compress, Bls12_381_G1_uncompress
+  , Bls12_381_G2_add, Bls12_381_G2_neg, Bls12_381_G2_scalarMul
+  , Bls12_381_G2_equal, Bls12_381_G2_hashToGroup
+  , Bls12_381_G2_compress, Bls12_381_G2_uncompress
+  , Bls12_381_millerLoop, Bls12_381_mulMlResult, Bls12_381_finalVerify
+  , Keccak_256, Blake2b_224
+  ]
+
+{- batch4b: IntegerToByteString and ByteStringToInteger.  These were enabled in
+ PlutusV3 at PV9, along with batch4a, They were enabled in PlutusV2 at PV10 in
+ #6056 and #6065.  They are available on the chain, but they're prohibitively
+ expensive because the proposal to update the relevant protocol parameters has
+ not (yet) been enacted.  This has left a "gap" in the cost model paramters: for
+ PlutusV3, the parameters for Batch 3 are followed those for 4a, then 4b, but
+ for PlutusV2 those for Batch3 are followed by those for Batch 4a, and those for
+ 4b aren't in use yet.  Since you can't actually use the 4b builtins in PlutusV2
+ at the moment, it's tempting to insert the 4a parameter before the 4b
+ parameters and enable them all at PV11 and with a suitable parameter update.
+ However, if we do do this there's a theoretical risk of turning a phase 2
+ failure into a phase 1 failure: would that be problematic?
+-}
+batch4b :: [DefaultFun]
+batch4b =
+  [ IntegerToByteString, ByteStringToInteger ]
+
+batch4 :: [DefaultFun]
+batch4 = batch4a ++ batch4b
+
+batch5 :: [DefaultFun]
+batch5 =
+  [ AndByteString, OrByteString, XorByteString, ComplementByteString
+  , ReadBit, WriteBits, ReplicateByte
+  , ShiftByteString, RotateByteString, CountSetBits, FindFirstSetBit
+  , Ripemd_160
+  ]
+
+batch6 :: [DefaultFun]
+batch6 =
+  [ ExpModInteger, DropList
+  , LengthOfArray, ListToArray, IndexArray
+  ]
+
+{-| Given a ledger language, return a map indicating which builtin functions were
+  introduced in which 'MajorProtocolVersion'.  This __must__ be updated when new
+  builtins are added.  It is not necessary to add entries for protocol versions
+  where no new builtins are added.  See Note [New builtins/language versions and
+  protocol versions]
 -}
+builtinsIntroducedIn :: PlutusLedgerLanguage -> Map.Map MajorProtocolVersion (Set.Set DefaultFun)
```

Comment:
This (and other functions in this module) are used to perform validity checks during on-chain deserialisation, so it's important that they're efficient.  I think that things like `batch2 ++ batch3 ++ batch4 ++ batch5 ++ batch6` will be computed at compile time, but `Set.fromList` and `Map.fromList` won't, so this isn't as efficient as it could be.  It seems that this problem can be [solved](https://well-typed.com/blog/2020/06/th-for-static-data/) using Template Haskell, and I plan to look into this in a later PR.

Reply:
I had to rework these functions because the previous versions relied on assumptions that are no longer valid (specifically, the predicate in `takeWhileAntitone` must satisfy a [certain property](https://hackage-content.haskell.org/package/containers-0.8/docs/Data-Map-Internal.html#v:takeWhileAntitone), and enabling new builtins in PlutusV1/V2 meant that this no longer held in the previous version of the function.

Reply:
Does this cause any performance regression? Previously `builtinsIntroducedIn` was a top-level Map that can be cached, but it is now a function.